Solved #3823 File annotation

upupming opened the pull request:
This pull request solved #3823, and improved the new line remover.

# Removed hyphens and newlines of highlighted text

I removed all new lines which **don't have preceded `.` or `:` becaues I think `.` or `:` is often used to start a new paragraph.

Test text:

<img src="https://raw.githubusercontent.com/upupming/paste/master/img/test-selection.png" width=300/>

Test result:

<img src="https://raw.githubusercontent.com/upupming/paste/master/img/20180802201505.png"/>

There is another problem with the regular expression `(\r?\n|\r)` when I try to use `(?<![.|:])(\r?\n|\r)` for excluding the new line end with `.` or `:`. 

If it's on Windows, the `\r\n` will be replaced with a whitespace, and it is OK. However if we encounter `.\r\n`, the `.\r` will be reserved and the `\n` will be replaced by a whitespace. In this way, the `.\r` will not be shown as new line int the field, and we don't need this whitespace. So instead we should change `.\r\n` to `\r\n` directly.

So I use java's `%n` to distinguish new line wisely.

# Font size of summary on the left changed to 10px

# Summary text wrapped

# The Final example:

<img src="https://raw.githubusercontent.com/upupming/paste/master/img/changed-file-notation.png"/>


<!-- describe the changes you have made here: what, why, ... 
     Link issues by using the following pattern: [#333](https://github.com/JabRef/jabref/issues/333) or [koppor#49](https://github.com/koppor/jabref/issues/47).
     The title of the PR must not reference an issue, because GitHub does not support autolinking there. -->


----

- [x] Change in CHANGELOG.md described
- [x] Tests created for changes
- [x] Manually tested changed features in running JabRef
- [x] Screenshots added in PR description (for bigger UI changes)
- [x] Ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

tobiasdiez commented on src/main/java/org/jabref/logic/formatter/bibtexfields/RemoveHyphenatedNewlinesFormatter.java:
```diff
@@ -10,7 +10,8 @@
 public class RemoveHyphenatedNewlinesFormatter extends Formatter {
-    private static final Pattern HYPHENATED_WORDS = Pattern.compile("(-\r\n|-\n|-\r)");
+    private static final String newLine = String.format("%n");
+    private static final Pattern HYPHENATED_WORDS = Pattern.compile("(-" + newLine + ")");
```
You can also use `\R` to match any line breaks, see https://docs.oracle.com/javase/9/docs/api/java/util/regex/Pattern.html.

upupming replied:
Thanks, it is what I didn't know before. But it seems `\R` will match any of `\r`, `\n` and `\r\n`.
![image](https://user-images.githubusercontent.com/24741764/43588283-bdeaa01a-969e-11e8-9c47-b9a5b27ebb1f.png)
In this example, you can see `\r` is matched on Windows. But we need only match `\r\n` on Windows, so I think `%n` is the right choice.

tobiasdiez commented on CHANGELOG.md:
```diff
@@ -27,6 +27,7 @@ We refer to [GitHub issues](https://github.com/JabRef/jabref/issues) by using `#
 - Window state is saved on close and restored on start.
 - Files without a defined external file type are now directly opened with the default aplication of the operating system
 - We streamlined the process to rename and move files by removing the confirmation dialogs.
+- We removed the redundant new lines of marking and wrapped the summary in File annotation tab.
```
Slightly better: We removed the redundant new lines of marking*s* and wrapped the summary in *the* File annotation tab.

Please also add a reference to the issue #3823 (cf. e.g. some of the changelog entries under "fixed").

upupming replied:
Fixed.

tobiasdiez commented on src/main/java/org/jabref/gui/entryeditor/fileannotationtab/FileAnnotationTabView.java:
```diff
@@ -94,12 +94,14 @@ private Node createFileAnnotationNode(FileAnnotationViewModel annotation) {
         Label page = new Label(Localization.lang("Page") + ": " + annotation.getPage());
 
-        marking.setStyle("-fx-font-weight: bold");
+        marking.setStyle("-fx-font-size: 10px; -fx-font-weight: bold");
```
Please specify the font-size in `em` so that it scales properly if the base font size is changed (e.g. on high res displays).

upupming replied:
Changed to `0.75em`

tobiasdiez commented on src/main/java/org/jabref/gui/entryeditor/fileannotationtab/FileAnnotationViewModel.java:
```diff
@@ -31,7 +31,11 @@ private void setupContentProperties(FileAnnotation annotation) {
-            this.marking.set(annotationContent.isEmpty() ? illegibleTextMessage : annotationContent);
+            String markingContent = (annotationContent.isEmpty() ? illegibleTextMessage : annotationContent);
+            // remove newlines && hyphens before linebreaks
+            markingContent = new RemoveHyphenatedNewlinesFormatter().format(markingContent);
```
It is good that you try to reuse existing logic, i.e. the existing formatters. However, these are also used to generate the bibtex key (and file name). For these use cases, it is still desired to remove all line breaks. Thus, I would prefer if you define the modified regex pattern to remove the linebreaks here in this class and leave the RemoveNewlinesFormatter untouched (except changing the pattern to use `\R`).

upupming replied:
Now I leave the two formatter untouched and only use the pattern in this scope.

tobiasdiez commented on src/main/java/org/jabref/gui/entryeditor/fileannotationtab/FileAnnotationViewModel.java:
```diff
@@ -31,7 +31,11 @@ private void setupContentProperties(FileAnnotation annotation) {
+            // remove newlines && hyphens before linebreaks
+            markingContent = new RemoveHyphenatedNewlinesFormatter().format(markingContent);
+            markingContent = new RemoveNewlinesFormatter().format(markingContent);
+            this.marking.set(markingContent);
```
It would be nice if you could add some tests verifying the new behavior: create a `FileAnnotion` object by hand (not by parsing a pdf) which contains such problematic hyphens/newlines and check that they are removed, indeed.

upupming replied:
Thank you for your advice, I have added a unit test!

upupming commented:
@tobiasdiez Thanks for your review! I will fix them soon.

upupming commented:
Hi, I reverted the change for `RemoveHyphenatedNewlinesFormatter` and `RemoveNewlinesFormatter`. I have also fixed the tooltip for marking summary.

**Before**
<img src="https://raw.githubusercontent.com/upupming/paste/master/img/20180803091540.png"/>
**After**
<img src="https://raw.githubusercontent.com/upupming/paste/master/img/20180803092209.png"/>

Now I'm going to add some unit tests!

upupming commented:
All done~ 😃

tobiasdiez commented on src/test/java/org/jabref/logic/formatter/bibtexfields/RemoveHyphenatedNewlinesFormatterTest.java:
```diff
@@ -15,8 +15,7 @@ public void setUp() {
-        assertEquals("water", formatter.format("wa-\nter"));
-        assertEquals("water", formatter.format("wa-\r\nter"));
-        assertEquals("water", formatter.format("wa-\rter"));
+        String newLine = String.format("%n");
```
Please keep these tests (here and below). The text that we want to format might contain linux line endings on Windows and vice versa. But you can of course also add a tests for ̀ %n`. 
Please change also all regex patterns to use `\R` instead of `%n` for newlines.

upupming replied:
Finally leaving them untouched. 

In summary, Pattern.compile("...") changed from `(\r?\n|\r)` to `\\R`

tobiasdiez commented:
Thanks for the quick follow-up! Looking forward to your next PR ;)
